doodles/Doodles.sol

```diff
@@ -0,0 +1,107 @@
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
+
+contract Doodles is ERC721, ERC721Enumerable, Ownable {
+    string public PROVENANCE;
+    bool public saleIsActive = false;
+    string private _baseURIextended;
+
+    bool public isAllowListActive = false;
+    uint256 public constant MAX_SUPPLY = 10000;
+    uint256 public constant MAX_PUBLIC_MINT = 5;
+    uint256 public constant PRICE_PER_TOKEN = 0.123 ether;
+
+    mapping(address => uint8) private _allowList;
+
+    constructor() ERC721("Doodles", "DOODLE") {
+    }
+	
+	// ## only the owner of this contract and execute this. isAllowListActive is a feature flag to allow/disallow adding the the _allowList in the mintAllowList function
+    function setIsAllowListActive(bool _isAllowListActive) external onlyOwner {
+        isAllowListActive = _isAllowListActive;
+    }
+
+    // ## only the owner call call this. This seems to be used to init the allow list AND limit the number of times an individual wallet can mint
+    function setAllowList(address[] calldata addresses, uint8 numAllowedToMint) external onlyOwner {
+        for (uint256 i = 0; i < addresses.length; i++) {
+            _allowList[addresses[i]] = numAllowedToMint;
+        }
+    }
+
+    // ## this is a read only function that anyone can execute. Since it is a view it requires no gas.  Returns the amount of items you can mint if you are white listed
+    function numAvailableToMint(address addr) external view returns (uint8) {
+        return _allowList[addr];
+    }
+
+    // ## allows folks to mint as long as they are on the allow list and meet someother basic critera
+    function mintAllowList(uint8 numberOfTokens) external payable {
+        uint256 ts = totalSupply();
+        require(isAllowListActive, "Allow list is not active");
+        require(numberOfTokens <= _allowList[msg.sender], "Exceeded max available to purchase");
+        require(ts + numberOfTokens <= MAX_SUPPLY, "Purchase would exceed max tokens");
+        require(PRICE_PER_TOKEN * numberOfTokens <= msg.value, "Ether value sent is not correct");
+
+        // ## decrement the amount of mints the sender can do
+        _allowList[msg.sender] -= numberOfTokens;
+
+        // ## loop through the number of tokens
+        for (uint256 i = 0; i < numberOfTokens; i++) {
+            _safeMint(msg.sender, ts + i);
+        }
+    }
+
+    function _beforeTokenTransfer(address from, address to, uint256 tokenId) internal override(ERC721, ERC721Enumerable) {
+        super._beforeTokenTransfer(from, to, tokenId);
+    }
+
+    function supportsInterface(bytes4 interfaceId) public view virtual override(ERC721, ERC721Enumerable) returns (bool) {
+        return super.supportsInterface(interfaceId);
+    }
+
+    function setBaseURI(string memory baseURI_) external onlyOwner() {
+        _baseURIextended = baseURI_;
+    }
+
+    function _baseURI() internal view virtual override returns (string memory) {
```

Comment:
This function is used on line 95 of ERC721.sol to concat the `baseURI` + `tokenId`

Reply:
TokenURI is called by external apps like Open Sea that want to display / import the NFT

Reply:
Thank you! I think I was a lil confused cause the tutorial I was going through was storing stuff on chain and was explicitly saving the token uris since the token uris contained the actual metadata including the image

This is a totally different case where after digging I see:
* the owner --> token id is stored in a map
* that token id will have a tokenUri that points to the metadata. Ex: https://ipfs.io/ipfs/QmPMc4tcBsMqLRuCQtPmPe84bpSjrC3Ky7t3JWuHXYB4aS/1
* * that metadata will have the image location in it. Ex: https://ipfs.io/ipfs/QmTDxnzcvj2p3xBrKcGv1wxoyhAn2yzCQnZZ9LmFjReuH9